tick/optim/solver/src/saga.h

```diff
@@ -0,0 +1,77 @@
+#ifndef TICK_OPTIM_SOLVER_SRC_SAGA_H_
+#define TICK_OPTIM_SOLVER_SRC_SAGA_H_
+
+// License: BSD 3 clause
+
+#include "array.h"
+#include "sgd.h"
+#include "../../prox/src/prox.h"
+#include "../../prox/src/prox_separable.h"
+
+class SAGA : public StoSolver {
+ public:
+  enum class VarianceReductionMethod {
+    Last = 1,
+    Average = 2,
+    Random = 3,
+  };
+
+ private:
+  double step;
+  // Probabilistic correction of the step-sizes of all model weights,
+  // given by the inverse proportion of non-zero entries in each feature column
+  ArrayDouble steps_correction;
+
+  VarianceReductionMethod variance_reduction;
+
+  ArrayDouble next_iterate;
+
+  bool solver_ready;
+  ArrayDouble gradients_memory;
```

Comment:
Maybe should we mention that this can be stored in an ArrayDouble because we are only applying SAGA on GLM